src/cortex_cli/token_manager.py

```diff
@@ -20,70 +20,156 @@
     """
-    with daemon.DaemonContext(stderr=open(errfile, 'w', encoding='UTF-8')):
-        start_token_manager(timeout, config)
+    with open(logfile, 'w', encoding='UTF-8') as output:
+        with daemon.DaemonContext(stdout=output, stderr=output):
+            start_token_manager(cycle, config)
 
 
-def start_token_manager(timeout: int, config: dict, single_run: bool = False) -> None:
+def start_token_manager(cycle: int, config: dict, single_run: bool = False) -> None:
     """Refresh tokens periodically.
+
+    For each refresh cycle new tokens are requested from auth server.
+    - If refresh is successful next refresh is attempted in the next cycle.
+    - If auth server does not respond refresh is attempted repeatedly until it succeeds or
+      the existing refresh token expires.
+    - If auth server responds but returns an error code or invalid tokens token manager is stopped.
+
     Args:
-        timeout: refresh timeout (period) in seconds
+        cycle: refresh cycle in seconds
         config: Cortex CLI configuration dict
         single_run: if True, refresh tokens only once and exit; otherwise repeat refreshing indefinitely
     """
-    path_to_tokens_dir = Path(config['tokens_file']).parent
-    path_to_tokens_file = config['tokens_file']
-    auth_server_url = config['auth_server_url']
-    realm = config['realm']
-    client_id = config['client_id']
 
     while True:
-        with open(path_to_tokens_file, 'r', encoding='utf-8') as file:
-            refresh_token = json.load(file)['refresh_token']
-
-        tokens = refresh_request(auth_server_url, realm, client_id, refresh_token)
-        if not tokens:
-            raise ClientAuthenticationError('Failed to update tokens. Probably, they were expired.')
-
-        timestamp = datetime.now()
-        tokens_json = json.dumps({
-            'pid': os.getpid(),
-            'timestamp': timestamp.isoformat(),
-            'access_token': tokens['access_token'],
-            'refresh_token': tokens['refresh_token'],
-            'auth_server_url': auth_server_url
-        })
-
-        try:
-            path_to_tokens_dir.mkdir(parents=True, exist_ok=True)
-            with open(Path(path_to_tokens_file), 'w', encoding='UTF-8') as file:
-                file.write(tokens_json)
-        except OSError as error:
-            print('Error writing tokens file', error)
+        tokens = read_tokens(config['tokens_file'])
+
+        new_tokens, status, sleep_time = refresh_tokens(config, tokens, cycle)
+        if new_tokens is None:
+            break
+
+        write_tokens(config['tokens_file'], config['auth_server_url'], status, **new_tokens)
 
         if single_run:
             break
 
-        human_timestamp = timestamp.strftime('%m/%d/%Y %H:%M:%S')
-        print(f'{human_timestamp}: Tokens refreshed successfully.')
-        time.sleep(timeout)
+        time.sleep(sleep_time)
+
+    print(f'{datetime.now().strftime("%m/%d/%Y %H:%M:%S")}: Token manager stopped')
```

Comment:
should this be `click.echo` instead of `print`?

Reply:
When token manager is run in daemon mode, the output is redirected to a file. When it is running with `--no-daemon` the output is printed out in the shell window. `Click.echo` would work well when output is a shell window, but `print` works nicely in both cases. I was only following existing convention here.

Reply:
Currently token manager is not dependent on click at all.

Reply:
Yeah, it can be done with the `logging` module. But let's think about it another time.